include/Mat.h

```diff
@@ -1278,33 +1238,51 @@ template<class Type2>
 void Mat<Type>::ito(Mat<Type2> &indices, Mat<Type> &out,
                 typename std::enable_if<std::is_integral<Type2>::value>::type*){
     errorCheck(indices.ndims != 1,
-        "Index list should be 1 dimension");
+        "index list should be 1 dimension");
     errorCheck(out.ndims != ndims,
-        "inconsistent number of dimensions in output matrix in call to ito");
-    errorCheck(ndims == 1 && out.columns() != indices.size(),
-        "output matrix shape mismatch - incorrect number of columns");
-    if(ndims > 1){
-        errorCheck(out.columns() != columns(),
-            "output matrix shape mismatch - incorrect number of columns");
-        errorCheck(out.rows() != indices.size(),
-            "output matrix shape mismatch - incorrect number of rows");
-    }
-    if(ndims == 2){
-        for(size_type i = 0; i < indices.size(); i++){
-            for(size_type j = 0; j < columns(); j++){
-                out(i,j) = operator()(indices(i),j);
+        "inconsistent number of dimensions in output matrix in call to ito()");
+    errorCheck(out.dims[0] != indices.size(),
+        "output matrix shape does not match given index list in call to ito()");
+    for(long i = 1; i < ndims; i++){
+        errorCheck(out.dims[i] != dims[i],
+        "output matrix shape mismatch in call to ito()");
+    }
+
+    size_type offset = size() / dims[0];
+    iterator dimend = begin();
+    iterator i = begin(); //for iterating the current matrix
+    size_type j = 0; //for iterating the index list
+    for(iterator k = out.begin(); k != out.end(); ++k){ //for iterating the output
```

Comment:
1. pull index loop out to outer loop:
   i=whatever
   k=whatever
   for(j=0;j<indices.size();j++)
   for(;i!=dimend;++i)
   ... *k=*i
   ++k;
2. inner conditional can be mostly removed